boot/seal.go

```diff
@@ -83,3 +86,102 @@ func sealKeyToModeenv(key secboot.EncryptionKey, model *asserts.Model, modeenv *
 
 	return nil
 }
+
+type bootChain struct {
+	Model          string      `json:"model"`
+	BrandID        string      `json:"brand-id"`
+	Grade          string      `json:"grade"`
+	ModelSignKeyID string      `json:"model-sign-key-id"`
+	AssetChain     []bootAsset `json:"asset-chain"`
+	Kernel         string      `json:"kernel"`
+	// KernelRevision is the revision of the kernel snap. It is empty if
+	// kernel is unasserted, in which case always reseal.
+	KernelRevision string `json:"kernel-revision"`
+	KernelCmdline  string `json:"kernel-cmdline"`
+}
+
+type bootAsset struct {
+	Role   string   `json:"role"`
+	Name   string   `json:"name"`
+	Hashes []string `json:"hashes"`
+}
+
+// helper types
+type sortedHashesBootAsset bootAsset
+type sortedAssetsBootChain bootChain
+
+// copyForMarshalling copies the contents of boot asset and sorts the hash list.
+func (b *bootAsset) copyForMarshalling() *sortedHashesBootAsset {
+	newB := *b
+	newB.Hashes = make([]string, len(b.Hashes))
+	copy(newB.Hashes, b.Hashes)
+	sort.Strings(newB.Hashes)
+	return (*sortedHashesBootAsset)(&newB)
+}
+
+// MarshalJSON marshals the boot asset into a form that is deterministic and
+// suitable for equivalence tests.
+func (b *bootAsset) MarshalJSON() ([]byte, error) {
+	s := (*sortedHashesBootAsset)(b.copyForMarshalling())
+	return json.Marshal(s)
+}
+
+type byBootAssetOrder []bootAsset
+
+func (b byBootAssetOrder) Len() int      { return len(b) }
+func (b byBootAssetOrder) Swap(i, j int) { b[i], b[j] = b[j], b[i] }
+func (b byBootAssetOrder) Less(i, j int) bool {
+	// sort order: role -> name -> hash list (len -> lexical)
+	byRole := b[i].Role < b[j].Role
+	byName := b[i].Name < b[j].Name
+	if b[i].Role != b[j].Role {
+		return byRole
+	}
+	if b[i].Name != b[j].Name {
+		return byName
+	}
+	return lessByHashList(b[i].Hashes, b[j].Hashes)
+}
+
+func lessByHashList(h1, h2 []string) bool {
+	if len(h1) != len(h2) {
+		return len(h1) < len(h2)
+	}
+	for idx := range h1 {
+		if h1[idx] < h2[idx] {
+			return true
+		}
+	}
+	return false
+}
+
+// copyForMarshalling copies the contents of bootChain and sorts boot assets.
+func (b *bootChain) copyForMarshalling() *sortedAssetsBootChain {
+	newB := *b
+	newB.AssetChain = make([]bootAsset, len(b.AssetChain))
+	for i := range b.AssetChain {
+		newB.AssetChain[i] = bootAsset(*b.AssetChain[i].copyForMarshalling())
+	}
+	return (*sortedAssetsBootChain)(&newB)
+}
+
+// MarshalJSON marshals the boot chain into a form that is deterministic and
+// suitable for equivalence tests.
+func (b *bootChain) MarshalJSON() ([]byte, error) {
+	s := b.copyForMarshalling()
+	sort.Sort(byBootAssetOrder(s.AssetChain))
+	return json.Marshal(s)
+}
+
+// equal returns true when boot chains are equivalent for reseal.
+func (b *bootChain) equalForReseal(other *bootChain) bool {
```

Comment:
I think we need sort on  []bootChain  and equalForReasl(bc1, bc2 []bootChain) more than this one.

We might need a comparison on single bootChains maybe for converting to the params, but I think it will need to be more specific than returning bool or not.  We kind of need to know where they are diverging in a useful way.